pkg/dataobj/streams_reader.go

```diff
@@ -0,0 +1,248 @@
+package dataobj
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/filemd"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/result"
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/sections/streams"
+)
+
+// Predicates for reading streams.
+type (
+	// LabelMatcher is a predicate for matching labels in a streams section.
+	// LabelMatcher predicates assert that a label named Name exists and its
+	// value is set to Value.
+	//
+	// For equality matches, LabelMatcher should always be used; LabelMatchers
+	// can translate into more efficient filter operations than a [LabelFilter]
+	// can.
+	LabelMatcher struct{ Name, Value string }
+
+	// LabelFilter is a predicate for matching labels in a streams section.
+	// LabelFilter predicates return a true value when the combination of the
+	// provided label name and value should be included in the result.
+	//
+	// LabelFilter predicates should be only used for more complex filtering; for
+	// equality matches, [LabelMatcher]s are more efficient.
+	LabelFilter func(name, value string) bool
+)
+
+// A Stream is an individual stream in a data object.
+type Stream struct {
```

Comment:
Not required now but some index queries like Volume and shard will require to get page statistics I think to be able to answer correctly.